README.md
# HTTP Request Context

[![npm package](https://nodei.co/npm/http-request-context.png?downloads=true&downloadRank=true&stars=true)](https://www.npmjs.com/package/http-request-context)

[![Build Status](https://travis-ci.org/zhujun24/http-request-context.svg)](https://travis-ci.org/zhujun24/http-request-context)
[![Coverage Status](https://coveralls.io/repos/github/zhujun24/http-request-context/badge.svg?branch=master)](https://coveralls.io/github/zhujun24/http-request-context?branch=master)
[![Known Vulnerabilities](https://snyk.io//test/github/zhujun24/http-request-context/badge.svg?targetFile=package.json)](https://snyk.io//test/github/zhujun24/http-request-context?targetFile=package.json)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fzhujun24%2Fhttp-request-context.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fzhujun24%2Fhttp-request-context?ref=badge_shield)
[![npm package](https://img.shields.io/npm/v/http-request-context.svg)](https://www.npmjs.com/package/http-request-context)

[![NPM downloads](https://img.shields.io/npm/dm/http-request-context.svg)](https://www.npmjs.com/package/http-request-context)
[![Dependency Status](https://david-dm.org/zhujun24/http-request-context.svg)](https://www.npmjs.com/package/http-request-context)
[![Dependency Status](https://david-dm.org/zhujun24/http-request-context/dev-status.svg)](https://www.npmjs.com/package/http-request-context)
[![Standard - JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://www.npmjs.com/package/http-request-context)

Set and Get request-scoped context anywhere.

## Requirement

### Nodejs version >= 8.2.0

This module uses the newer [async_hooks](https://nodejs.org/api/async_hooks.html#async_hooks_async_hooks) API which is considered `Experimental` by Nodejs.

## Options

| Option | Description | Type | Default |
|:------------|:------------|:------------|:------------|
| interval | remove expired callstack interval(s) | Number | 10
| expire | callstack expire time(s)| Number | 150
| removeAfterFinish | remove callstack after [http.ServerResponse](https://nodejs.org/api/http.html#http_class_http_serverresponse) [finish](https://nodejs.org/api/http.html#http_event_finish) | Boolean | false
| removeAfterClose | remove callstack after [http.ServerResponse](https://nodejs.org/api/http.html#http_class_http_serverresponse) [close](https://nodejs.org/api/http.html#http_event_close_1) | Boolean | false

#### options.interval

Remove expired callstack interval, used like `setInterval(removeExpiredCallstack, interval)`.

#### options.expire

Callstack expire time, must be longer than full lifecycle of a request.

#### options.removeAfterFinish

It will actively remove the relevant callstack after [http.ServerResponse](https://nodejs.org/api/http.html#http_class_http_serverresponse) [finish](https://nodejs.org/api/http.html#http_event_finish).

If set to `true`, you can get the context synchronously in the finish event, but not asynchronous. The benefit is that it can improve the performance of this middleware.

#### options.removeAfterClose

This is very similar to `options.removeAfterFinish`, the difference is that after the close event.

Please Note! if set to `true`, in some cases, the close event may be caused by the client terminating the request, after the close event, we may still use the context after the incomplete asynchronous operation is completed, this will result in loss of context.

## Init Middleware

**This module is recommend as a "top-level" middleware for ensure all context can be tracked**

- `httpRequestContext.middleware(options)` Init Express middleware.
- `httpRequestContext.koaMiddleware(options)` Init Koa middleware.

## Set Context

- `httpRequestContext.set(key, value)` Set context anywhere.
- `httpRequestContext.set({ key: value })` This is also OK.

## Get Context

- `httpRequestContext.get(key)` Get the `[key]` attribute of the context.
- `httpRequestContext.get()` Gets an object containing all context properties.

## How to Use

see [example](https://github.com/zhujun24/http-request-context/tree/master/example) here.

### Install

```npm
npm install http-request-context --save
```

### Express

#### Init

```js
import httpRequestContext from 'http-request-context'

app.use(httpRequestContext.middleware())
```

#### Set Context

```js
import httpRequestContext from 'http-request-context'

// set context by key-value
app.use((req, res, next) => {
  setTimeout(() => {
    httpRequestContext.set('foo', 'bar')
    next()
  }, 100)
})
```

#### Get Context

```js
import httpRequestContext from 'http-request-context'

httpRequestContext.get('foo') // 'bar'
```

### Koa

#### Init

```js
import httpRequestContext from 'http-request-context'

app.use(httpRequestContext.koaMiddleware())
```

#### Set Context

```js
import httpRequestContext from 'http-request-context'

// set context by key-value
app.use(async (ctx, next) => {
  await new Promise(resolve => {
    setTimeout(() => {
      httpRequestContext.set('user', 'user')
      resolve()
    }, 300)
  })
  await next()
})
```

##### Get Context

```js
import httpRequestContext from 'http-request-context'

httpRequestContext.get('foo') // 'bar'
```

## Lost Context Tips

#### http.ServerResponse close event

Sometimes, when client terminate request by close window or reload page, it will cause http.ServerResponse emit 'close' event, this event is trigger by root, so it break away from current request scope, in this case, we can add `res`(express) or `ctx.res`(koa) parameter to get context function to ensure context can be tracked, as follows:

```js
// Express
res.on('close', () => {
  console.log('close', httpRequestContext.get('foo', res))
})

// Koa
ctx.res.on('close', () => {
  console.log('close', httpRequestContext.get('foo', ctx.res))
})
```

#### MySQL

If you init mysql connect before http server start, you may get context undefined in mysql query callback scope.

[googleapis/cloud-trace-nodejs #946](https://github.com/googleapis/cloud-trace-nodejs/issues/946)

[nodejs/node #22360](https://github.com/nodejs/node/issues/22360)

```js
mysqlConnection.query('SELECT * FROM table', (error, results, fields) => {
  httpRequestContext.get('foo') // undefined
})
```

You can use util.promisify to avoid it.

```js
util.promisify(mysqlConnection.query).bind(mysqlConnection)('SELECT * FROM table')
  .then((results, fields) => {
    httpRequestContext.get('foo') // 'bar'
  })
  .catch(error => {})
```


History.md
0.3.1 / 2020-04-22
==================

**fixes**
  * 🐛 [[`9a4fe20`](https://github.com/zhujun24/http-request-context/commit/9a4fe204c0e5776bca10e7231c0add24375bb535)] fixup rootID init

0.3.0 / 2019-10-22
==================

**features**
  * ✨ [[`4fc9eea`](https://github.com/zhujun24/http-request-context/commit/4fc9eea957247222656be4617b080755254eba6c)] add removeAfterFinish&removeAfterClose option

0.2.5 / 2019-10-19
==================

**fixes**
  * 🐛 [[`dbbf561`](https://github.com/zhujun24/http-request-context/commit/dbbf56169b1cf2ca55507ad7de46ead9d63b5af1)]  remove destroy callstack to fix context lost

0.1.0 / 2019-10-19
==================

**release**
  * 🔖 [[`2eb4a0e`](https://github.com/zhujun24/http-request-context/commit/2eb4a0efeb6e7b4291aa29ea9657361eb64220cd)]  Release first version!


test/common.js
module.exports = {
  port: 3003,

  sleep: (time = 300) => new Promise(resolve => {
    setTimeout(resolve, time)
  })
}


test/koa.test.js
/* eslint-disable no-undef */

const chai = require('chai')
const chaiHttp = require('chai-http')
const assert = require('assert')
const Koa = require('koa')
const httpRequestContext = require('../index')
const { port } = require('./common')

const koaApp = new Koa()

chai.use(chaiHttp)

describe('Koa Middleware Test', async function () {
  it('Koa return 200 and correct value', async function () {
    koaApp.use(httpRequestContext.koaMiddleware({
      interval: 0.2,
      expire: 0.1
    }))

    koaApp.use(async (ctx, next) => {
      await new Promise(resolve => {
        setTimeout(() => {
          httpRequestContext.set('key', 'value')
          resolve()
        }, 50)
      })
      await next()
    })

    koaApp.use(ctx => {
      ctx.body = httpRequestContext.get('key')
    })

    const server = koaApp.listen(port)

    const response = await chai.request(server).get('/')

    server.close()

    assert.strictEqual(200, response.status)
    assert.strictEqual('value', response.text)
  })
})


test/index.test.js
/* eslint-disable no-undef */

process.env.HTTP_REQUEST_CONTEXT_INTERVAL = 200
process.env.HTTP_REQUEST_CONTEXT_EXPIRE = 100

const chai = require('chai')
const chaiHttp = require('chai-http')
const assert = require('assert')
const express = require('express')
const Koa = require('koa')
const httpRequestContext = require('../')

const expressPort = 3004
const koaPort = 3005

const expressApp = express()
const koaApp = new Koa()

// Ensure timer execution and increase coverage
const sleep = () => new Promise(resolve => {
  // large than interval and timeout
  setTimeout(resolve, 300)
})

chai.use(chaiHttp)

describe('Express Middleware Test', function () {
  it('Express return 200 and correct value', async function () {
    expressApp.use(httpRequestContext.middleware)

    expressApp.use((req, res, next) => {
      setTimeout(function () {
        httpRequestContext.set('key', 'value')
        httpRequestContext.set({ improve: 'coverage' }) // improve set func coverage
        next()
      }, 50)
    })

    expressApp.use(function (req, res) {
      httpRequestContext.get() // improve get func coverage
      res.send(httpRequestContext.get('key'))
    })

    const server = expressApp.listen(expressPort, function () {
      console.log(`Express is listening at port ${expressPort}`)
    })

    await sleep()

    const response = await chai.request(server).get('/')

    server.close()

    assert.strictEqual(200, response.status)
    assert.strictEqual('value', response.text)
  })
})

describe('Koa Middleware Test', async function () {
  it('Koa return 200 and correct value', async function () {
    koaApp.use(httpRequestContext.koaMiddleware)

    koaApp.use(async (ctx, next) => {
      await new Promise(resolve => {
        setTimeout(() => {
          httpRequestContext.set('key', 'value')
          resolve()
        }, 50)
      })
      await next()
    })

    koaApp.use(ctx => {
      ctx.body = httpRequestContext.get('key')
    })

    const server = koaApp.listen(koaPort, function () {
      console.log(`Koa is listening at port ${koaPort}`)
    })

    await sleep()

    const response = await chai.request(server).get('/')

    server.close()

    assert.strictEqual(200, response.status)
    assert.strictEqual('value', response.text)
  })
})


index.js
const asyncHooks = require('async_hooks')

const config = {}
const callstackMap = new Map() // all callstack map
const TCPWrapCallstackContainers = new Map() // request root callstack
const DEFAULT_INTERVAL = 10 // remove expired context interval
const DEFAULT_EXPIRE = 150 // callstack expire time (must be longer than full lifecycle of a request)
const RES_ROOT_ID = '__HTTP_REQUEST_CONTEXT_ROOT_ID__'

// remove callstack after request finish/close
const removeCallstackTrigger = rootId => {
  if (TCPWrapCallstackContainers.has(rootId)) {
    TCPWrapCallstackContainers.get(rootId).id.forEach(asyncId => {
      callstackMap.delete(asyncId)
    })
    TCPWrapCallstackContainers.delete(rootId)
  }
}

// remove asyncId interval
const interval = () => {
  setTimeout(interval, config.interval)

  const now = Date.now()

  for (const asyncId of callstackMap.keys()) {
    if (now - callstackMap.get(asyncId).__tm < config.expire) {
      break
    } else {
      callstackMap.delete(asyncId)
    }
  }

  for (const rootId of TCPWrapCallstackContainers.keys()) {
    if (now - TCPWrapCallstackContainers.get(rootId).__tm < config.expire) {
      break
    } else {
      removeCallstackTrigger(rootId)
    }
  }

  callstackMap.delete(asyncHooks.executionAsyncId())
}

// find callstack root
const findRootId = id => {
  const callstack = callstackMap.get(id)
  if (callstack) {
    if (callstack.data) {
      return id
    }
    return findRootId(callstack.id)
  }
}

asyncHooks.createHook({
  init (asyncId) {
    callstackMap.set(asyncId, {
      id: asyncHooks.executionAsyncId(),
      __tm: Date.now()
    })

    if (config.optimize) {
      const rootId = findRootId(asyncId)
      if (rootId && TCPWrapCallstackContainers.has(rootId)) {
        TCPWrapCallstackContainers.get(rootId).id.push(asyncId)
      }
    }
  }
}).enable()

const middleware = res => {
  const rootId = asyncHooks.executionAsyncId()
  res[RES_ROOT_ID] = rootId

  callstackMap.set(rootId, {
    id: rootId,
    data: {},
    __tm: Date.now()
  })

  if (config.optimize) {
    TCPWrapCallstackContainers.set(rootId, {
      id: [rootId],
      __tm: Date.now()
    })
  }

  if (config.removeAfterFinish) {
    res.on('finish', () => {
      process.nextTick(() => {
        removeCallstackTrigger(rootId)
      })
    })
  }

  if (config.removeAfterClose) {
    res.on('close', () => {
      process.nextTick(() => {
        removeCallstackTrigger(rootId)
      })
    })
  }
}

const initOptions = (options = {}) => {
  Object.assign(config, {
    interval: (options.interval || DEFAULT_INTERVAL) * 1000,
    expire: (options.expire || DEFAULT_EXPIRE) * 1000,
    removeAfterFinish: !!options.removeAfterFinish,
    removeAfterClose: !!options.removeAfterClose,
    optimize: !!options.removeAfterFinish || !!options.removeAfterClose
  })
}

module.exports = {
  middleware: options => {
    initOptions(options)
    interval()

    return (req, res, next) => {
      // res.set('connection', 'close')
      middleware(res)
      next()
    }
  },

  koaMiddleware: options => {
    initOptions(options)
    interval()

    return async (ctx, next) => {
      // ctx.set('connection', 'close')
      middleware(ctx.res)
      await next()
    }
  },

  set: (key, value) => {
    const rootId = findRootId(asyncHooks.executionAsyncId())
    if (rootId && callstackMap.has(rootId)) {
      const data = Object.prototype.toString.call(key) === '[object Object]' ? key : { [key]: value }
      Object.assign(callstackMap.get(rootId).data, data)
    }
  },

  get: (key, res) => {
    const rootId = res && res[RES_ROOT_ID] ? res[RES_ROOT_ID] : findRootId(asyncHooks.executionAsyncId())
    if (rootId && callstackMap.has(rootId)) {
      const { data } = callstackMap.get(rootId)
      return typeof key === 'undefined' ? data : data[key]
    }
  }
}


namespace.js
const asyncHooks = require('async_hooks')

let topAsyncId = -1

class NameSpace {
  constructor() {
    this.activeContext = null
    this.contextList = []
    this.contexts = new Map()
  }

  set(key, val) {
    if (!this.activeContext) {
      throw new Error('No active context available.')
    }
    this.activeContext[key] = val
    return true
  }

  get(key) {
    if (!this.activeContext) {
      return undefined
    }
    return this.activeContext[key]
  }

  createContext() {
    return this.activeContext || {}
  }

  init() {
    this.enter(this.createContext())
  }

  enter(context) {
    this.contextList.push(this.activeContext)
    this.activeContext = context
  }

  exit(context) {
    if (this.activeContext === context) {
      this.activeContext = this.contextList.pop()
      return
    }

    const index = this.contextList.lastIndexOf(context)
    if (~index) {
      this.contextList.splice(index, 1)
    }
  }
}

module.exports = () => {
  const namespace = new NameSpace();

  asyncHooks.createHook({
    init: (asyncId, type, triggerId, resource) => {
      topAsyncId = asyncHooks.executionAsyncId()

      if (namespace.activeContext) {
        namespace.contexts.set(asyncId, namespace.activeContext)
      } else if (topAsyncId === 0) {
        // CurrentId will be 0 when triggered from C++. Promise events
        // https://github.com/nodejs/node/blob/master/doc/api/async_hooks.md#triggerid
        const triggerId = asyncHooks.triggerAsyncId()
        const triggerIdContext = namespace.contexts.get(triggerId)
        if (triggerIdContext) {
          namespace.contexts.set(asyncId, triggerIdContext)
        }
      }
    },

    before: asyncId => {
      topAsyncId = asyncHooks.executionAsyncId()

      const context = namespace.contexts.get(asyncId) || namespace.contexts.get(topAsyncId)
      if (context) {
        namespace.enter(context)
      }
    },

    after: asyncId => {
      topAsyncId = asyncHooks.executionAsyncId()

      const context = namespace.contexts.get(asyncId) || namespace.contexts.get(topAsyncId)
      if (context) {
        namespace.exit(context)
      }
    },

    destroy: asyncId => {
      topAsyncId = asyncHooks.executionAsyncId()

      namespace.contexts.delete(asyncId)
    }
  })
    .enable()

  return namespace
}


example/koa.js
const util = require('util')
const Koa = require('koa')
const httpRequestContext = require('../')

const app = new Koa()

app.use(httpRequestContext.koaMiddleware({
  removeAfterClose: true,
  removeAfterFinish: true
}))

app.use(async (ctx, next) => {
  httpRequestContext.set('name', 'zhujun24')
  ctx.res.on('close', () => {
    console.log('close', httpRequestContext.get('name', ctx.res))
  })
  ctx.res.on('finish', () => {
    console.log('finish', httpRequestContext.get('name'))
  })
  await next()
})

app.use(async (ctx, next) => {
  await new Promise(resolve => {
    setTimeout(() => {
      httpRequestContext.set('sex', 'male')
      resolve()
    }, 300)
  })
  await next()
})

app.use(async (ctx, next) => {
  await util.promisify(process.nextTick)().then(() => {
    httpRequestContext.set('age', '24')
  })
  await next()
})

app.use(ctx => {
  ctx.body = {
    name: httpRequestContext.get('name'),
    sex: httpRequestContext.get('sex'),
    age: httpRequestContext.get('age')
  }
})

app.listen(3002)


example/express.js
const util = require('util')
const express = require('express')
const httpRequestContext = require('../')

const app = express()

app.use(httpRequestContext.middleware({
  removeAfterClose: true,
  removeAfterFinish: true
}))

app.use((req, res, next) => {
  httpRequestContext.set('name', 'zhujun24')
  res.on('close', () => {
    console.log('close', httpRequestContext.get('name', res))
  })
  res.on('finish', () => {
    console.log('finish', httpRequestContext.get('name'))
  })
  next()
})

app.use((req, res, next) => {
  setTimeout(() => {
    httpRequestContext.set('sex', 'male')
    next()
  }, 300)
})

app.use((req, res, next) => {
  util.promisify(process.nextTick)().then(() => {
    httpRequestContext.set('age', '24')
    next()
  })
})

app.all('*', (req, res) => {
  res.send({
    name: httpRequestContext.get('name'),
    sex: httpRequestContext.get('sex'),
    age: httpRequestContext.get('age')
  })
})

app.listen(3001)
